### ISSUE TITLE
Leverage Django REST Framework

### ISSUE THREAD

**Issue by cancan101** [jonashagstedt/swampdragon#11]
Given [this description](http://swampdragon.net/tutorial/part-1-here-be-dragons-and-thats-a-good-thing/):

> Model serializers
> Because SwampDragon will be communicating with the frontend via JavaScript, and JavaScript don't understand Django models, we need to tell SwampDragon how to serialize these models into JSON (short for JavaScript Object Notation).

why not use Django REST Framework (DRF)'s serializers rather than re-inventing the wheel? Code [here](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/serializers.py) and description [here](http://www.django-rest-framework.org/api-guide/serializers/) and [here](http://www.django-rest-framework.org/tutorial/1-serialization/). Perhaps use the [renderers](http://www.django-rest-framework.org/api-guide/renderers/)? They seem to solve the exact goal that you outlined.

I know you mentioned that you did not use them here: https://news.ycombinator.com/item?id=8676967:

> About the serializers: You do need separate serializers from DRF (they are not the same package after all).

but you don't address the reason why.

**Comment by cancan101** [comment]
You should loop in @tomchristie but I think now that DRF is on v3, the API for Serializers should be relatively stable.

**Comment by jonashagstedt** [comment]
It would be nice if DRF serializers could be used.

I think a solution where the existing serializers and DRF serializers can both work would be ideal.

If I rewrote the way serializers are handled this could be a possibility (even go as far as making serializers swappable).

Whatever solution I (or someone else) come up with, should not create a direct dependency on DRF 3.x
and it should be backwards compatible with the current version of SD. 

I am considering this though. 
The SD serializers are not without their shortcomings and there are things I want to change with them so I might plan out an update and see if I can include a way to use the DRF serializers as well.

**Comment by tomchristie** [comment]
More than happy to help out with any questions if helpful.

I wouldn't be too concerned about tying yourself to API stability with REST framework - the 3.0 upgrade was something of an exception - for medium point releases (ie 3.x - which is what we'll be having for the foreseeable future) we have a strong deprecation policy that's essentially in line with what Django does.

http://www.django-rest-framework.org/topics/release-notes/#deprecation-policy

Also I can't see the serializers changing substantially at this point.

**Comment by pcompassion** [comment]
+1!

**Comment by jonashagstedt** [comment]
Quick update on this:
TL;DR: It will be done but it's going to take some considerable time

I've been poking at this for a bit now and the serializers are very different indeed, and this is going to be quite an arduous task.

At this point I'm more inclined to rewrite both the serializers and route handlers which is not going to be a small task.

The DRF serializers are (of course) not aware of the data maps.

What this actually means:

The data map contains information on how objects should be mapped.


	class Parent(models.Model):
		name = models.CharField(max_length=100)


	class Child(models.Model):
		name = models.CharField(max_length=100)
		parent = models.ForeignKey(Parent)


If you have a serializer for `Parent` including children, when I make a change to a `Child` it will contain information for the data mapper to know which parent it belongs to (rather than serializing EVERYTHING every time, this means a small message can be sent to the client with just the updated fields and the mapping data) 

This doesn't only affect the route handler but the self publishing models as well.

In short: I'm happy to do this, but the work it involves is not tiny so it's going to take quite some time to do this (fitting this around haivng a baby and my day job).

**Comment by jonashagstedt** [comment]
This is now available in the 0.4.3 branch (it's experimental at the moment)

**Comment by silentninja** [comment]
Was waiting for this update!!!!.Really excited to test it out

**Comment by cho-is** [comment]
+1!

**Comment by jonashagstedt** [comment]
If anyone has tried the 0.4.3 branch I would love some feedback

**Comment by sylflo** [comment]
I'm trying to do the first tutorial. And after something a little more complex.
In DRF you can use a model to create a serializer. But SwampDragon needs a serializer in its model.

I've got serializers.py

```
from rest_framework import serializers
from .models import TodoList, TodoItem

class TodoListSerializer(serializers.ModelSerializer):
    class Meta:
        model = TodoList
        fields = ('id', 'name', 'description')
        publish_fields = ('name', 'description')


class TodoItemSerializer(serializers.ModelSerializer):
    class Meta:
        model = TodoItem
        fields = ('id', 'done', 'text')
        publish_fields = ('done', 'text')
        update_fields = ('done', )
```

And models.py

```
class TodoList(SelfPublishModel, models.Model):
    serializer_class = TodoListSerializer
    name = models.CharField(max_length=100)
    description = models.TextField()


class TodoItem(SelfPublishModel, models.Model):
    serializer_class = TodoItemSerializer
    todo_list = models.ForeignKey(TodoList)
    done = models.BooleanField(default=False)
    text = models.CharField(max_length=100)

```


And I got
```
ImportError: cannot import name TodoList
```

I don't know how to resolve this issue.
I want to use model = in the serializer file if possible.

**Comment by silentninja** [comment]
@sylflo Please create a new issue.
Now answering the question, its a circular import error.
##### Solution:Use a string of the model name in the metaclass of the serializer instead of importing the actual model.

**Comment by pcompassion** [comment]
Would it be possible to define multiple serializers for one model class?


`SwampDragon needs a serializer in its model.` makes me ask the question.

**Comment by jonashagstedt** [comment]
You can create several serializers for one model but you can not have multiple serializers on a selfpublish model. However I am currently unsure of the future of the self publishing model. People don't seem to be able to use it properly and get upset when it's either creating too many queries or not enough information so at the time of writing this I would say don't use the self publish model and let the routers handle the publishing

**Comment by pcompassion** [comment]
@jonashagstedt Thanks for the quick answer. I'll try the tutorial and will be back to you asap.

**Comment by jsenecal** [comment]
@jonashagstedt do you have any updates on this? I am myself a strong DRF user and just stumbled across your project (it looks awesome btw) but would like to avoid having to write another set of serializers.

**Comment by jonashagstedt** [comment]
@jsenecal You could try the 0.4.3 branch.
I would love some help with feedback on this
